File: app/src/main/java/com/anthonysottile/kenken/settings/StatisticsManager.java
package com.anthonysottile.kenken.settings;

import android.content.SharedPreferences;

import com.anthonysottile.kenken.ui.UIConstants;

import org.json.JSONArray;
import org.json.JSONException;

public final class StatisticsManager {

    private static final String Statistics = "Statistics";
    // TODO: implement dirty cache methodology

    private static SharedPreferences preferences = null;
    private static GameStatistics[] statistics = null;

    private static void saveGameStatistics() {
        // Create JSON array
        JSONArray array = new JSONArray();
        try {
            for (int i = 0; i < UIConstants.GameSizes; i += 1) {
                array.put(i, StatisticsManager.statistics[i].ToJson());
            }
        } catch (JSONException e) {
            e.printStackTrace();
        }

        SharedPreferences.Editor editor =
                StatisticsManager.preferences.edit();

        editor.putString(StatisticsManager.Statistics, array.toString());
        editor.apply();
    }

    private static void loadStatistics() {
        // First check if we have the Statistics already there
        if (!StatisticsManager.preferences.contains(StatisticsManager.Statistics)) {
            StatisticsManager.ClearGameStatistics();
        } else {
            StatisticsManager.statistics = new GameStatistics[UIConstants.GameSizes];

            try {
                JSONArray array =
                        new JSONArray(
                                StatisticsManager.preferences.getString(
                                        StatisticsManager.Statistics,
                                        ""
                                )
                        );

                for (int i = 0; i < UIConstants.GameSizes; i += 1) {
                    StatisticsManager.statistics[i] = new GameStatistics(array.getJSONObject(i));
                }

            } catch (JSONException e) {
                e.printStackTrace();
            }
        }
    }

    /**
     * Clears the statistics.
     */
    public static void ClearGameStatistics() {

        StatisticsManager.statistics = new GameStatistics[UIConstants.GameSizes];
        for (int i = 0; i < UIConstants.GameSizes; i += 1) {
            StatisticsManager.statistics[i] = new GameStatistics(i + UIConstants.MinGameSize);
        }

        StatisticsManager.saveGameStatistics();
    }

    /**
     * Returns the Game Statistics for the specified game size.
     *
     * @param gameSize The game size to retrieve statistics for.
     * @return The Game Statistics for the specified game size.
     */
    public static GameStatistics GetGameStatistics(int gameSize) {
        return StatisticsManager.statistics[gameSize - UIConstants.MinGameSize];
    }

    /**
     * Call when a game has been started to update that statistic.
     *
     * @param gameSize The size of the game started.
     */
    public static void GameStarted(int gameSize) {
        int index = gameSize - UIConstants.MinGameSize;
        StatisticsManager.statistics[index].GameStarted();

        StatisticsManager.saveGameStatistics();
    }

    /**
     * Call when a game has ended to update statistics.
     * Returns true if this is a new high score.
     *
     * @param gameSize The size of the game completed.
     * @param ticks    The number of ms that the game took.
     * @return True if the game is a new high score.
     */
    public static boolean GameEnded(int gameSize, long ticks) {

        // Returns true if the game is a high score
        boolean returnValue = false;

        int index = gameSize - UIConstants.MinGameSize;
        GameStatistics game = StatisticsManager.statistics[index];

        int gameSeconds = (int) (ticks / 1000);

        game.GameWon(gameSeconds);

        if (gameSeconds < game.getBestTime()) {
            game.SetBestTime(gameSeconds);
            returnValue = true;
        }

        StatisticsManager.saveGameStatistics();

        return returnValue;
    }

    /**
     * Initialization method to give the manager a reference to preferences.
     *
     * @param preferences A reference to the application preferences.
     */
    public static void Initialize(SharedPreferences preferences) {
        StatisticsManager.preferences = preferences;
        StatisticsManager.loadStatistics();
    }

    private StatisticsManager() {
    }
}

File: app/src/main/java/com/anthonysottile/kenken/settings/StatisticsManager.kt
package com.anthonysottile.kenken.settings

import android.content.SharedPreferences

import com.anthonysottile.kenken.ui.UIConstants

import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

object StatisticsManager {
    private const val Statistics = "Statistics"
    private const val HardStatistics = "HardStatistics"
    // TODO: implement dirty cache methodology

    private lateinit var preferences: SharedPreferences
    private lateinit var statistics: Array<GameStatistics>
    private lateinit var hardStatistics: Array<GameStatistics>

    private fun gameSizeToIndex(gameSize: Int): Int {
        return gameSize - UIConstants.MinGameSize
    }

    private fun statsJson(): JSONArray {
        val statsArray = JSONArray()
        this.statistics.forEach { statsArray.put(it.toJson()) }
        return statsArray
    }

    private fun hardStatsJson(): JSONArray {
        val hardStatsArray = JSONArray()
        this.hardStatistics.forEach { hardStatsArray.put(it.toJson()) }
        return hardStatsArray
    }

    private fun saveGameStatistics() {
        val editor = this.preferences.edit()
        editor.putString(this.Statistics, this.statsJson().toString())
        editor.putString(this.HardStatistics, this.hardStatsJson().toString())
        editor.apply()
    }

    private fun loadStatistics() {
        if (this.preferences.contains(this.Statistics)) {
            try {
                val arr = JSONArray(
                        this.preferences.getString(StatisticsManager.Statistics, "")
                )
                this.statistics = Array(UIConstants.GameSizes) { i ->
                    GameStatistics(arr.getJSONObject(i))
                }
            } catch (e: JSONException) {
                this.clearStatistics()
            }
        } else {
            this.clearStatistics()

        }
        if (this.preferences.contains(this.HardStatistics)) {
            try {
                val arr = JSONArray(
                        this.preferences.getString(StatisticsManager.HardStatistics, "")
                )
                this.hardStatistics = Array(UIConstants.GameSizes) { i ->
                    GameStatistics(arr.getJSONObject(i))
                }
            } catch (e: JSONException) {
                this.clearHardStatistics()
            }
        } else {
            this.clearHardStatistics()
        }
    }

    private fun clearStatistics() {
        this.statistics = Array(UIConstants.GameSizes) { i ->
            GameStatistics(i + UIConstants.MinGameSize)
        }
    }

    private fun clearHardStatistics() {
        this.hardStatistics = Array(UIConstants.GameSizes) { i ->
            GameStatistics(i + UIConstants.MinGameSize)
        }
    }

    fun exportStatistics(): String {
        val obj = JSONObject()
        obj.put("stats", this.statsJson())
        obj.put("hardStats", this.hardStatsJson())
        return obj.toString()
    }

    fun importStatistics(statsString: String): Boolean {
        try {
            val obj = JSONObject(statsString)
            val statsJson = obj.getJSONArray("stats")
            val stats = Array(UIConstants.GameSizes) { i ->
                GameStatistics(statsJson.getJSONObject(i))
            }
            val hardStatsJson = obj.getJSONArray("hardStats")
            val hardStats = Array(UIConstants.GameSizes) { i ->
                GameStatistics(hardStatsJson.getJSONObject(i))
            }

            this.statistics = stats
            this.hardStatistics = hardStats
            this.saveGameStatistics()

            return true
        } catch (e: JSONException) {
            e.printStackTrace()
            return false
        }
    }

    fun clearAllStatistics() {
        this.clearStatistics()
        this.clearHardStatistics()
        this.saveGameStatistics()
    }

    fun getGameStatistics(gameSize: Int, hardMode: Boolean): GameStatistics {
        val index = this.gameSizeToIndex(gameSize)
        val arr = if (hardMode) this.hardStatistics else this.statistics
        return arr[index]
    }

    fun gameStarted(gameSize: Int, hardMode: Boolean) {
        this.getGameStatistics(gameSize, hardMode).gameStarted()
        this.saveGameStatistics()
    }

    /**
     * Call when a game has ended to update statistics.
     * Returns true if this is a new high score.
     *
     * @param gameSize The size of the game completed.
     * @param ticks    The number of ms that the game took.
     * @return True if the game is a new high score.
     */
    fun gameEnded(gameSize: Int, hardMode: Boolean, ticks: Long): Boolean {
        val gameSeconds = (ticks / 1000).toInt()
        val highScore = this.getGameStatistics(gameSize, hardMode).gameWon(gameSeconds)

        this.saveGameStatistics()

        return highScore
    }

    fun initialize(preferences: SharedPreferences) {
        this.preferences = preferences
        this.loadStatistics()
    }
}
